### PR TITLE
frontmatter fixes; plan modifier packages

### PR DESCRIPTION
**Pull request by joshbeard**
* Move the 'frontmatter' code to its own package for better organization
  and importability across sub-packages.
* Create a new `otherattributemodifier` package to modify an attribute
  when _another_ attribute changes.
  * Fixes bug that causes "inconsistent plan" errors on docs. This was
    caused by the "updated_at" and "revision" attributes continuously
    changing when other attributes changed. These attributes should
    change when most writable attributes are changed.
  * Makes the plan modifier more robust and usable across the data types
    that need it.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -38,8 +38,8 @@ lint: modverify vet gofumpt lines golangci-lint ## Run all linters
 ## Testing ##
 .PHONY: test
 test: ## Run unit and race tests with 'go test'
-	go test -v -count=1 -parallel=4 -coverprofile=coverage.txt -covermode count ./readme/...
-	go test -race -short ./readme/...
```
Comment (joshbeard) [inline Makefile:42]: Don't need to recurse. The sub-packages are tested with the parent tests.

File: readme/custom_page_resource.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/liveoaklabs/readme-api-go-client/readme"
+	"github.com/liveoaklabs/terraform-provider-readme/readme/frontmatter"
```
Comment (joshbeard) [inline readme/custom_page_resource.go:15]: Move `frontmatter.go` to new package for better organization

File: readme/doc_resource.go [hunk 3]
```diff
@@ -416,6 +418,23 @@ func (r *docResource) Schema(
...
+							otherattributemodifier.StringModifyString(path.Root("parent_doc"), "ParentDoc", true),
+							otherattributemodifier.StringModifyString(
+								path.Root("parent_doc_slug"),
+								"ParentDocSlug",
+								true,
+							),
+							otherattributemodifier.StringModifyString(path.Root("slug"), "Slug", true),
+							otherattributemodifier.StringModifyString(path.Root("title"), "Title", true),
+							otherattributemodifier.StringModifyString(path.Root("type"), "Type", true),
+						},
```
Comment (joshbeard) [inline readme/doc_resource.go:437]: This section configures the `updated_at` parameter to get updated when any of the listed attributes here change. Due to the typing, this is broken out into separate functions based on the source type and the "other" attribute's type.

File: readme/doc_test.go [hunk 4]
```diff
@@ -106,9 +106,6 @@ var mockDoc readme.Doc = readme.Doc{
 	Version:       mockVersion.ID,
 }
 
-// mockDocBodyString represents a "raw" body string provided in the configuration with extraneous whitespace.
-var mockDocBodyString string = fmt.Sprintf(`  \n%s\n\n `, mockDoc.Body)
-
```
Comment (joshbeard) [inline readme/doc_test.go:111]: No longer needed

File: readme/frontmatter/frontmatter.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+// package frontmatter includes types and functions that provide resource plan
```
Comment (joshbeard) [inline readme/frontmatter/frontmatter.go:1]: This is the contents of the existing `frontmatter.go` copied into its own package for improved re-usability and organization.

File: readme/otherattributemodifier/bool.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
+package otherattributemodifier
```
Comment (joshbeard) [inline readme/otherattributemodifier/bool.go:1]: New package to enable triggering a change of an attribute value when another attribute value changes. These are needed on some of the unstable attributes provided by the ReadMe API, such as `updated_at` and `revision`.

These enable those attributes to get updated at the same time as the 'other' attribute. This addresses a bug that caused inconsistent state after applying changes in some resources (e.g. docs).

File: readme/provider.go [hunk 7]
```diff
@@ -209,52 +208,6 @@ func (p *readmeProvider) Resources(_ context.Context) []func() resource.Resource
 	}
 }
 
-// otherAttributeModifier is a plan modifier that plans a change for an attribute if another specified attribute is changed.
```
Comment (joshbeard) [inline readme/provider.go:212]: Replaced by the new "otherattributemodifier" package

**Review by bdwellons — APPROVED** [review]

**Pull request merged**